{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# TCGA: Pre-processing DNA mutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Imports\n",
    "import matplotlib.pyplot as plt\n",
    "import networkx as nx\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import seaborn as sb\n",
    "import sklearn as sk\n",
    "import statsmodels as sm\n",
    "import sys\n",
    "import os\n",
    "\n",
    "from pathlib import Path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Some path utilities\n",
    "def ls(path):\n",
    "    return [p for p in path.iterdir()]\n",
    "\n",
    "\n",
    "def find_files(path, pattern=None):\n",
    "    \"\"\" Recursively find all files \"\"\"\n",
    "    all_files = list()\n",
    "    if path.is_dir():\n",
    "        all_files.extend([f for p in ls(path) for f in find_files(p)])\n",
    "    else:\n",
    "        all_files.append(path)\n",
    "    if pattern is not None:\n",
    "        all_files = [f for f in all_files if f.match(pattern)]\n",
    "    return all_files\n",
    "\n",
    "# OK, let's do some duck-typing\n",
    "Path.ls = ls\n",
    "Path.find_files = find_files\n",
    "\n",
    "# Which variants are expected to be 'high impact'?\n",
    "high_impact_variants = set(['Frame_Shift_Del', 'Frame_Shift_Ins', 'In_Frame_Del'\n",
    "                   , 'In_Frame_Ins', 'Indel', 'Intron', 'Missense'\n",
    "                   , 'Missense_Mutation', 'Nonsense_Mutation', 'Nonstop_Mutation'\n",
    "                   , 'Read-through', 'Splice_Site', 'Splice_Site_Del'\n",
    "                   , 'Splice_Site_Ins', 'Splice_Site_SNP', 'Translation_Start_Site'])\n",
    "\n",
    "\n",
    "def path_to_sample(path):\n",
    "    \"\"\" Return the sample name from MAF file path \"\"\"\n",
    "    return path.stem.split('.')[0]\n",
    "\n",
    "# A simple test case\n",
    "assert 'TCGA-IH-A3EA-01' == path_to_sample(Path('data/tcga/gdac.broadinstitute.org_SKCM.Mutation_Packager_Calls.Level_3.2016012800.0.0/TCGA-IH-A3EA-01.maf.txt'))\n",
    "\n",
    "\n",
    "def path_to_cancer_type(path):\n",
    "    \"\"\" Return the cancre type MAF file path \n",
    "    For example, path='data/tcga/gdac.broadinstitute.org_SKCM.Mutation_Packager_Calls.Level_3.2016012800.0.0/TCGA-IH-A3EA-01.maf.txt'\n",
    "    we want to return 'SKCM'\n",
    "    \"\"\"\n",
    "    return path.parent.stem.split('.')[2].split('_')[1]\n",
    "\n",
    "# A simple test case\n",
    "assert 'SKCM' == path_to_cancer_type(Path('data/tcga/gdac.broadinstitute.org_SKCM.Mutation_Packager_Calls.Level_3.2016012800.0.0/TCGA-IH-A3EA-01.maf.txt'))\n",
    "\n",
    "\n",
    "def mutated_genes(maf_file):\n",
    "    \"\"\" Parse the MAF file and return a dictionary of genes with\n",
    "    the count of 'high impact' mutations per gene \"\"\"\n",
    "    # print(f\"Mutated genes: '{maf_file}'\")\n",
    "    try:\n",
    "        df = pd.read_csv(maf_file, sep='\\t', low_memory=False)\n",
    "    except UnicodeDecodeError:\n",
    "        print(f\"ERROR reading file '{maf_file}'\")\n",
    "        return None\n",
    "    keep_rows = [vc in high_impact_variants for vc in df.Variant_Classification]\n",
    "    count_by_gene = dict()\n",
    "    for gene in df[keep_rows].Hugo_Symbol:\n",
    "        count_by_gene[gene] = count_by_gene.get(gene, 0) + 1\n",
    "    return count_by_gene\n",
    "\n",
    "\n",
    "def process_maf_dir(path):\n",
    "    \"\"\" Find all MAF files in 'path' and get a dictionary of\n",
    "    mutated genes per cancer_type and sample \"\"\"\n",
    "    by_cancer_type = dict()\n",
    "    for p in path.find_files('*maf.txt'):\n",
    "        cancer_type = path_to_cancer_type(p)\n",
    "        if cancer_type not in by_cancer_type:\n",
    "            print(f\"Adding cancer type '{cancer_type}'\")\n",
    "            by_cancer_type[cancer_type] = dict()\n",
    "        sample = path_to_sample(p)\n",
    "        by_cancer_type[cancer_type][sample] = mutated_genes(p)\n",
    "    return by_cancer_type\n",
    "\n",
    "\n",
    "def get_all_genes(by_sample):\n",
    "    \"\"\" Return a (sorted) list of genes in all samples \"\"\"\n",
    "    all_genes = set([g for gene_dict in by_sample.values() if gene_dict is not None for g in gene_dict.keys()])\n",
    "    all_genes = list(all_genes)\n",
    "    all_genes.sort()\n",
    "    return all_genes\n",
    "\n",
    "\n",
    "def variants_df(by_sample):\n",
    "    \"\"\" Create a dataframe containing number of variants per sample \"\"\"\n",
    "    df = pd.DataFrame(by_sample, index=get_all_genes(by_sample), columns=sorted(list(by_sample.keys())))\n",
    "    return df.fillna(0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load and pre-process DNA mutations\n",
    "Load all mutations, filter them and convert them in a single dataframe per cancer type"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "path = Path('data/tcga')\n",
    "sample_by_cancer_type = process_maf_dir(path)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Convert all "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Adding cancer type 'PAAD'\n",
      "Adding cancer type 'ACC'\n",
      "Adding cancer type 'LUAD'\n",
      "Adding cancer type 'READ'\n",
      "Adding cancer type 'LUSC'\n",
      "Adding cancer type 'ESCA'\n",
      "Adding cancer type 'GBMLGG'\n",
      "Adding cancer type 'STAD'\n",
      "Adding cancer type 'OV'\n",
      "Adding cancer type 'DLBC'\n",
      "Adding cancer type 'LGG'\n",
      "Adding cancer type 'UCEC'\n",
      "Adding cancer type 'PRAD'\n",
      "Adding cancer type 'CESC'\n",
      "Adding cancer type 'COADREAD'\n",
      "Adding cancer type 'UVM'\n",
      "Adding cancer type 'GBM'\n",
      "Adding cancer type 'THYM'\n",
      "Adding cancer type 'PCPG'\n",
      "Adding cancer type 'STES'\n",
      "Adding cancer type 'KIRC'\n",
      "Adding cancer type 'UCS'\n",
      "Adding cancer type 'SKCM'\n",
      "ERROR reading file 'data/tcga/gdac.broadinstitute.org_SKCM.Mutation_Packager_Calls.Level_3.2016012800.0.0/TCGA-IH-A3EA-01.maf.txt'\n",
      "Adding cancer type 'BRCA'\n",
      "Adding cancer type 'TGCT'\n",
      "Adding cancer type 'CHOL'\n",
      "Adding cancer type 'LIHC'\n",
      "Adding cancer type 'HNSC'\n",
      "Adding cancer type 'LAML'\n",
      "Adding cancer type 'KIPAN'\n",
      "Adding cancer type 'COAD'\n",
      "Adding cancer type 'KICH'\n",
      "Adding cancer type 'SARC'\n",
      "Adding cancer type 'BLCA'\n",
      "Adding cancer type 'THCA'\n",
      "Adding cancer type 'KIRP'\n",
      "Cancer type: PAAD\n",
      "Cancer type: ACC\n",
      "Cancer type: LUAD\n",
      "Cancer type: READ\n",
      "Cancer type: LUSC\n",
      "Cancer type: ESCA\n",
      "Cancer type: GBMLGG\n",
      "Cancer type: STAD\n",
      "Cancer type: OV\n",
      "Cancer type: DLBC\n",
      "Cancer type: LGG\n",
      "Cancer type: UCEC\n",
      "Cancer type: PRAD\n",
      "Cancer type: CESC\n",
      "Cancer type: COADREAD\n",
      "Cancer type: UVM\n",
      "Cancer type: GBM\n",
      "Cancer type: THYM\n",
      "Cancer type: PCPG\n",
      "Cancer type: STES\n",
      "Cancer type: KIRC\n",
      "Cancer type: UCS\n",
      "Cancer type: SKCM\n",
      "Cancer type: BRCA\n",
      "Cancer type: TGCT\n",
      "Cancer type: CHOL\n",
      "Cancer type: LIHC\n",
      "Cancer type: HNSC\n",
      "Cancer type: LAML\n",
      "Cancer type: KIPAN\n",
      "Cancer type: COAD\n",
      "Cancer type: KICH\n",
      "Cancer type: SARC\n",
      "Cancer type: BLCA\n",
      "Cancer type: THCA\n",
      "Cancer type: KIRP\n"
     ]
    }
   ],
   "source": [
    "df_by_cancer_type = dict()\n",
    "for cancer_type in sample_by_cancer_type.keys():\n",
    "    print(f\"Cancer type: {cancer_type}\")\n",
    "    by_sample = sample_by_cancer_type[cancer_type]\n",
    "    df_by_cancer_type[cancer_type] = variants_df(sample_by_cancer_type[cancer_type])\n",
    "    file = f\"{path}/{cancer_type}.mutations_by_gene.csv\"\n",
    "    print(f\"Saving to file '{file}'\")\n",
    "    df_by_cancer_type[cancer_type].to_csv(file)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "mlgen",
   "language": "python",
   "name": "mlgen"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
